### PR TITLE
feat: User can follow other users

### PR DESCRIPTION
**Pull request by jojitoon**
#### What does this PR do?
Implement functionality to allow user to follow other users

#### Description of Task to be completed?

- modify database relationship
-  setup functionality to allow user follow other users
-  setup functionality to allow user unfollow other users
-  setup functionality to allow user view user following them or those they follow

#### How should this be manually tested?

- Clone the repository.
- Checkout the `ft-user-followings-159987409`  branch.
- Run npm install on your local machine.
- Run the db migrations using npm run migrate.
- Start the server using npm run start_dev.
- Using postman or any similar. get  a token upon registration and use it to test these routes:
  `localhost:8000/api/v1/users/:userId/follow`
`localhost:8000/api/v1/users/:userId/unfollow`
`localhost:8000/api/v1/users/followings`

#### Any background context you want to provide?
This feature will enable a user to follow an author or reader and also view them

#### What are the relevant pivotal tracker stories?
 #159987409

### CODE REVIEW

File: server/controllers/userController.js [hunk 1]
```diff
@@ -66,6 +66,100 @@ const userController = {
...
+  /**
+   * @description Method to unfollow a user
+  * @param  {object} req The request
+   * @param  {object} res The response
+   * @returns {object} json response
+   */
+  unfollow: (req, res) => {
+    Users
+      .findById(req.params.userId)
+      .then((user) => {
```
Comment (iAmao) [inline server/controllers/userController.js:189]: No catch block for this. If something happens users will never get a response

File: server/controllers/userController.js [hunk 2]
```diff
@@ -66,6 +66,100 @@ const userController = {
...
+          return res.status(400).jsend.fail({ message: 'you cannot unfollow yourself' });
+        }
+        Followings
+          .findOne({
+            where: {
+              followed: user.id,
+              follower: req.currentUser.id
+            }
+          })
+          .then((follow) => {
```
Comment (iAmao) [inline server/controllers/userController.js:206]: No catch block for this. If something happens users will never get a response

File: server/controllers/userController.js [hunk 3]
```diff
@@ -66,6 +66,100 @@ const userController = {
...
+  /**
+    * @description Method to follow a user
+    * @param  {object} req The request
+    * @param  {object} res The response
+    * @returns {object} json response
+    */
+  following: (req, res) => {
+    Users
+      .findOne({ where: { id: req.currentUser.id }, include: ['followed', 'follower'] })
+      .then((user) => {
```
Comment (iAmao) [inline server/controllers/userController.js:234]: No catch block for this. If something happens users will never get a response

**Review by iAmao — CHANGES_REQUESTED** [review]
Nice, but please ensure to always handle errors and edge cases.

File: test/test.spec.js [hunk 4]
```diff
@@ -253,3 +258,136 @@ describe('USER SIGN IN TEST', () => {
...
+      .request(app)
+      .post('/api/v1/users/1/follow')
+      .set('authorization', faketoken)
+      .end((err, res) => {
+        expect(res.body.status).to.equal('error');
+        expect(res.body.message).to.equal('Failed to authenticate token! Valid token required');
+        done();
+      });
+  });
+  it('should return user does not exist', (done) => {
```
Comment (iAmao) [inline test/test.spec.js]: All these tests have the same descriptions

**Review by iAmao — CHANGES_REQUESTED** [review]

File: server/controllers/userController.js [hunk 5]
```diff
@@ -131,7 +132,108 @@ const userController = {
         message: 'Success!',
         users,
         count: users.length
-      }));
+      }))
+      .catch(() => res.status(500).jsend.error('Login was not successful. Please try again'));
```
Comment (iAmao) [inline server/controllers/userController.js]: Be uniform in your error responses

File: server/controllers/userController.js [hunk 6]
```diff
@@ -131,7 +132,108 @@ const userController = {
...
+            }
+          })
+          .spread((_following, created) => {
+            // check if relationship exist
+            if (!created) {
+              return res.status(400).jsend.fail({ message: `You are already following ${user.username}` });
+            }
+            res.status(200).jsend.success({ message: `you are now following ${user.username}` });
+          })
+          .catch(() => res.status(500).jsend.error('You could not follow user at this time . Please try again'));
```
Comment (iAmao) [inline server/controllers/userController.js]: Be uniform in you error response

File: server/controllers/userController.js [hunk 7]
```diff
@@ -131,7 +132,108 @@ const userController = {
...
+      })
+      .catch(() => res.status(500).jsend.error('You could not follow user at this time . Please try again'));
```
Comment (iAmao) [inline server/controllers/userController.js]: Be uniform in you error response

File: server/controllers/userController.js [hunk 8]
```diff
@@ -131,7 +132,108 @@ const userController = {
...
+                message: `you are not following ${user.username}`,
+              });
+            }
+
+            follow
+              .destroy()
+              .then(() => res.status(200).jsend.success({
+                message: `you are no longer following ${user.username}`
+              }))
+              .catch(() => res.status(500).jsend.error('You could not unfollow user at this time . Please try again'));
```
Comment (iAmao) [inline server/controllers/userController.js]: Be uniform in you error response

File: server/controllers/userController.js [hunk 9]
```diff
@@ -131,7 +132,108 @@ const userController = {
...
+          })
+          .catch(() => res.status(500).jsend.error('You could not unfollow user at this time . Please try again'));
```
Comment (iAmao) [inline server/controllers/userController.js]: Be uniform in you error response

File: server/controllers/userController.js [hunk 10]
```diff
@@ -131,7 +132,108 @@ const userController = {
...
+      })
+      .catch(() => res.status(500).jsend.error('You could not unfollow user at this time . Please try again'));
```
Comment (iAmao) [inline server/controllers/userController.js]: Be uniform in you error response

File: server/controllers/userController.js [hunk 11]
```diff
@@ -131,7 +132,108 @@ const userController = {
...
+    * @returns {object} json response
+    */
+  following: (req, res) => {
+    Users
+      .findOne({ where: { id: req.currentUser.id }, include: ['followed', 'follower'] })
+      .then((user) => {
+        const { followed, follower } = user;
+        return res.status(200).jsend.success({ followed, follower });
+      })
+      .catch(() => res.status(500).jsend.error('we could not get you details at this time . Please try again'));
```
Comment (iAmao) [inline server/controllers/userController.js]: Be uniform in you error response

File: test/articles.spec.js [hunk 12]
```diff
@@ -23,6 +23,7 @@ describe('ARTICLE ENDPOINT TESTS', () => {
         password: 'Password'
       })
       .end((err, res) => {
+        console.log(res.body);
```
Comment (iAmao) [inline test/articles.spec.js]: remove logs

**Review by iAmao — CHANGES_REQUESTED** [review]

**Review by iAmao — APPROVED** [review]
